### ISSUE TITLE
bandwidth_test  is stuck when using random message size

### ISSUE THREAD

**Issue by LiPengze97** [Derecho-Project/derecho#187]
In the origin bandwidth_test, the program just sending fixed size of messages 
https://github.com/Derecho-Project/derecho/blob/b499931ef71142cef12c0289347266dbfaaf0bfc/src/applications/tests/performance_tests/bandwidth_test.cpp#L140
```
    auto send_all = [&]() {
        Replicated<RawObject>& raw_subgroup = group.get_subgroup<RawObject>();
        for(uint i = 0; i < num_messages; ++i) {
            // the lambda function writes the message contents into the provided memory buffer
            // in this case, we do not touch the memory region
            raw_subgroup.send(max_msg_size, [](char* buf) {});
        }
    };
```
when I try to use random size messages size like followings, the program won't stop:
```
    auto send_all = [&]() {
        Replicated<RawObject>& raw_subgroup = group.get_subgroup<RawObject>();
        for(uint i = 0; i < num_messages; ++i) {
            // the lambda function writes the message contents into the provided memory buffer
            // in this case, we do not touch the memory region
            size_t curr_size = rand() % max_msg_size;
            dbg_default_info("{} messages sent with size {}", i, curr_size);
            raw_subgroup.send(curr_size, [](char* buf) {});
        }
    };
```
the logs are
```
[02:55:33.959409] [derecho_debug] [Thread 4129] [info] 9999 messages sent with size 6550
[02:55:33.959426] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 856, num_received = 13691, sender rank = 0. Reading from SST row 0, slot 113003
[02:55:33.959448] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 857, num_received = 13701, sender rank = 1. Reading from SST row 1, slot 51365
[02:55:33.959459] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 856, num_received = 13692, sender rank = 0. Reading from SST row 0, slot 123276
[02:55:33.959482] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 857, num_received = 13702, sender rank = 1. Reading from SST row 1, slot 61638
[02:55:33.959504] [derecho_debug] [Thread 4163] [trace] Updating seq_num for subgroup 0 to 27385
[02:55:33.959544] [derecho_debug] [Thread 4129] [trace] Subgroup 0: get_sendbuffer_ptr increased future_message_indices to 13704
[02:55:33.959603] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 856, num_received = 13693, sender rank = 0. Reading from SST row 0, slot 133549
[02:55:33.959620] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 857, num_received = 13703, sender rank = 1. Reading from SST row 1, slot 71911
[02:55:33.959642] [derecho_debug] [Thread 4163] [trace] Updating seq_num for subgroup 0 to 27387
[02:55:33.959740] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 856, num_received = 13694, sender rank = 0. Reading from SST row 0, slot 143822
[02:55:33.959767] [derecho_debug] [Thread 4163] [trace] Updating seq_num for subgroup 0 to 27389
[02:55:33.959829] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 856, num_received = 13695, sender rank = 0. Reading from SST row 0, slot 154095
[02:55:33.959861] [derecho_debug] [Thread 4163] [trace] Updating seq_num for subgroup 0 to 27391
[02:55:33.959913] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 857, num_received = 13696, sender rank = 0. Reading from SST row 0, slot 0
[02:55:33.959948] [derecho_debug] [Thread 4163] [trace] Updating seq_num for subgroup 0 to 27393
[02:55:33.959969] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27376
[02:55:33.959978] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27377
[02:55:33.959996] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27378
[02:55:33.960004] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27379
[02:55:33.960032] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27380
[02:55:33.960040] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27381
[02:55:33.960056] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27382
[02:55:33.960064] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27383
[02:55:33.960076] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27384
[02:55:33.960083] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27385
[02:55:33.960095] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27386
[02:55:33.960103] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27387
[02:55:33.960115] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27388
[02:55:33.960122] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27389
[02:55:33.960134] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27390
[02:55:33.960142] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27391 and least_undelivered_seq_num=27391
[02:55:33.960497] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 857, num_received = 13697, sender rank = 0. Reading from SST row 0, slot 10273
[02:55:33.960530] [derecho_debug] [Thread 4163] [trace] Updating seq_num for subgroup 0 to 27395
[02:55:33.960625] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 857, num_received = 13698, sender rank = 0. Reading from SST row 0, slot 20546
[02:55:33.960656] [derecho_debug] [Thread 4163] [trace] Updating seq_num for subgroup 0 to 27397
[02:55:33.960752] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 857, num_received = 13699, sender rank = 0. Reading from SST row 0, slot 30819
[02:55:33.960783] [derecho_debug] [Thread 4163] [trace] Updating seq_num for subgroup 0 to 27399
[02:55:33.960857] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 857, num_received = 13700, sender rank = 0. Reading from SST row 0, slot 41092
[02:55:33.960889] [derecho_debug] [Thread 4163] [trace] Updating seq_num for subgroup 0 to 27401
[02:55:33.960969] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 857, num_received = 13701, sender rank = 0. Reading from SST row 0, slot 51365
[02:55:33.961000] [derecho_debug] [Thread 4163] [trace] Updating seq_num for subgroup 0 to 27403
[02:55:33.961090] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 857, num_received = 13702, sender rank = 0. Reading from SST row 0, slot 61638
[02:55:33.961122] [derecho_debug] [Thread 4163] [trace] Updating seq_num for subgroup 0 to 27405
[02:55:33.961218] [derecho_debug] [Thread 4163] [trace] receiver_trig calling sst_receive_handler_lambda. next_seq = 857, num_received = 13703, sender rank = 0. Reading from SST row 0, slot 71911
[02:55:33.961249] [derecho_debug] [Thread 4163] [trace] Updating seq_num for subgroup 0 to 27407
[02:55:33.961273] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27392
[02:55:33.961282] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27393
[02:55:33.961301] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27394
[02:55:33.961308] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27395
[02:55:33.961336] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27396
[02:55:33.961345] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27397
[02:55:33.961362] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27398
[02:55:33.961369] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27399
[02:55:33.961382] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27400
[02:55:33.961389] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27401
[02:55:33.961401] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27402
[02:55:33.961409] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27403
[02:55:33.961421] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27404
[02:55:33.961428] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27405
[02:55:33.961440] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27406
[02:55:33.961448] [derecho_debug] [Thread 4163] [trace] Subgroup 0, can deliver a locally stable SST message: min_stable_num=27407 and least_undelivered_seq_num=27407
```
In other words, the bool flag `done` will never be changed by `stability_callback`. It seems that the sending job can not be finished.
I tested the program with two procedures on the same machine. When I use fixed size message, the program will produce the `data_derecho_bw` with reasonable results. In the random message scenario, the program is stuck and does not produce any output.
Why can't Derecho support random size message sending ? Is it because I'm not using it properly?

**Comment by KenBirman** [comment]
This says it will send 9999 messages but in fact as coded you print that "Sending" line once for each message.  One thing I'm noticing is that rand() % max_message_size could easily do a send with 0 bytes.  I'm not sure you get a stability callback in that case.

**Comment by LiPengze97** [comment]
Yes, you are right, if the message size is fixed at 0, the program won't get a stability callback. If the message size is rand() % max_message_size + 1.
My fault, sorry.

**Comment by KenBirman** [comment]
Actually, perhaps that is a bug!  Why should the API behave differently for a message of size 0?  A null message gets a delivery slot, like any other.  Shouldn’t it do a stability callback, too?

**Comment by Panlichen** [comment]
In another trial with `bandwidth_test`, I set the `send_all` lambda as:
```c++
auto send_all = [&]() {
    Replicated<RawObject>& raw_subgroup = group.get_subgroup<RawObject>();
    for(uint i = 0; i < num_messages; ++i) {
        size_t curr_size = rand() % max_msg_size;
        if(curr_size == 0)
            dbg_default_info("{} message sent with size {}", i, curr_size);
        raw_subgroup.send(curr_size, [](char* buf) {});
    }
};
```
and the `stability_callback` lambda as:
```c++
auto stability_callback = [&done,
                            total_num_messages,
                            num_delivered = 0u](uint32_t subgroup,
                                                uint32_t sender_id,
                                                long long int index,
                                                std::optional<std::pair<char*, long long int>> data,
                                                persistent::version_t ver) mutable {
    ++num_delivered;
    dbg_default_warn("{}(total: {}) messages delivered", num_delivered, total_num_messages);
    if(num_delivered == total_num_messages) {
        done = true;
    }
};
```
then run `bandwidth_test 2 2 10000 0` on two nodes(recall: this cmd means 2 nodes, one_sender, 10000 messages, ordered mode), and I got stuck at 9997 messages. Here are some useful logs:
```shell
# on the sender side
...
[21:37:12.159611] [derecho_debug] [Thread 31218] [info] 1740 message sent with size 0
...
[21:37:12.159651] [derecho_debug] [Thread 31254] [warning] 1740(total: 10000) messages delivered
...
[21:37:12.204446] [derecho_debug] [Thread 31218] [info] 6139 message sent with size 0
...
[21:37:12.204507] [derecho_debug] [Thread 31254] [warning] 6139(total: 10000) messages delivered
...
[21:37:12.241817] [derecho_debug] [Thread 31254] [warning] 9996(total: 10000) messages delivered
[21:37:12.241821] [derecho_debug] [Thread 31254] [warning] 9997(total: 10000) messages delivered
# then stuck
```
```shell
# on the receiver side
...
[21:37:12.159648] [derecho_debug] [Thread 31255] [warning] 1740(total: 10000) messages delivered
...
[21:37:12.204502] [derecho_debug] [Thread 31255] [warning] 6139(total: 10000) messages delivered
...
[21:37:12.241815] [derecho_debug] [Thread 31255] [warning] 9996(total: 10000) messages delivered
[21:37:12.241816] [derecho_debug] [Thread 31255] [warning] 9997(total: 10000) messages delivered
# then stuck
```

**Comment by sagarjha** [comment]
> Actually, perhaps that is a bug! Why should the API behave differently for a message of size 0? A null message gets a delivery slot, like any other. Shouldn’t it do a stability callback, too?

@KenBirman It is by design, otherwise the application will be confused to see a callback for a message it did not send. Automated null messages have size 0 that are not sent by the application of course.

**Comment by Panlichen** [comment]
Thanks for your timely reaction, but the problem we were encountered is not solved. Originally we found that when using random size messages in `bandwith_test`, the program stuck. As I mentioned in the last comment, the message with size 0 is appropriately processed(at least the log shows so), the 9997th message stuck, and it has a non-zero size.

**Comment by KenBirman** [comment]
Sagar is saying that it is illegal to send a message of size 0.  Please modify your code to not have this error.  If the issue is reproduced again, even with no message of size 0, we can investigate further.

**Comment by Panlichen** [comment]
Sorry to have bothered you with that, if there is no message with size 0, the program runs smoothly.

**Comment by sagarjha** [comment]
You can of course send messages of size 0 from your application. But they will not result in a callback from Derecho. Just check the exit condition of bandwidth test. It counts up to a certain number of delivered messages. So you have to send at least that many non-zero sized messages. That's all this is about. Nothing too profound.

**Comment by Panlichen** [comment]
Thank you so much! In our experiment, we observed 3 messages with a size of 0 and ended up getting 9997 replies, that's exactly what you describe! Thank you!

**Comment by songweijia** [comment]
@Panlichen and @LiPengze97 sent email to me saying that eliminating the zero-sized messages does not fix the issue completely. I can reproduce it. I found the reason lying in `src/core/p2p_connection_manager.cpp:84`. To tell if an incoming message is a null reply, the code tested only the first byte in the messgae header, which is UNRELIABLE as the size type is `size_t`. With my fix in commit 1116d36, I can successfully use random sized messages. @Panlichen and @LiPengze97, plesae check if this works for you.

@xinzheyang  @sagarjha Please confirm the bugfix in commit 1116d36.

**Comment by sagarjha** [comment]
Hi Weijia, it looks good. Thanks for fixing the bug. I am closing the issue.
